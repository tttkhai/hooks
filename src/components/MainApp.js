import React, { useState, useCallback, useReducer } from "react";
import Hello from "./Hello";
import useFetch from "./useFetch";
import Bye from "./Bye";
import ListItem from './ListItem';


export const reducer = (state, action) => {
  switch (action.type) {
    case "add-todo":
      return {
        todo: [...state.todo, { data: action.payload, complete: false }],
      };
    case "toggle-todo":
      return {
        todo: state.todo.map((res, idx) =>
          action.payload === idx ? { ...res, complete: !res.complete } : res
        ),
      };
    default:
      return state.todo;
  }
};

export default function MainApp({ expenses }) {
  const [pageNumber, setPageNumber] = useState(0);
  const [count, setCount] = useState(0);
  const url = `https://randomuser.me/api?page=${pageNumber}`;
  const { data, loading } = useFetch(url);
  const [showHello, setShowHello] = useState(true);
  const [{ todo }, dispatch] = useReducer(reducer, { todo: [] });
  const [text, setText] = useState("");
  const increment = useCallback(
    (n) => {
      setCount((c) => c + n);
    },
    [setCount]
  );

  return (
    <div>
      <div>
        <button onClick={() => setPageNumber(pageNumber + 1)}> Click</button>
      </div>
      <div>{loading ? "loading..." : data}</div>
      <button onClick={() => setShowHello(!showHello)}> toggle</button>
      {showHello && <Hello />}
      <Bye increment={increment} />
      <div>{count}</div>
      <form
        onSubmit={(e) => {
          e.preventDefault();
          dispatch({ type: "add-todo", payload: text });
          setText("");
        }}
      >
        <input
          value={text}
          name="text"
          onChange={(e) => setText(e.target.value)}
        />
      </form>
      {todo.map((res, index) => {
        return (
          <div
            style={{ textDecoration: res.complete ? "line-through" : "" }}
            key={index}
            onClick={() => dispatch({ type: "toggle-todo", payload: index })}
          >
            {res.data}
          </div>
        );
      })}
      <div className="expenses">
        {expenses.length===0?  
          (<p>Nothing</p>):
          (expenses.map((res) => {
            return <ListItem key={res.id} className="hello" {...res} /> 
          }))
        }
      </div>
      {/* <input type="email" onChange={(e)=> setInput(e.target.value)}/> */}
    </div>
  );
}

// export default MainApp;
