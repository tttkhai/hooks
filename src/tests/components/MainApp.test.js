import React from "react";
import MainApp from "../../components/MainApp";
import { shallow } from "enzyme";
import expenses from "../fixtures/text";

test("should render MainApp dynamically", () => {
  const wrapper = shallow(
    <MainApp
      expenses={expenses}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test("test result at class expenses", () => {
    const wrapper = shallow(<MainApp expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});

test("should render MainApp with empty message", () => {
  const wrapper = shallow(<MainApp expenses={[]} />);
  expect(wrapper.find('.expenses p').text()).toBe('Nothing');
});

