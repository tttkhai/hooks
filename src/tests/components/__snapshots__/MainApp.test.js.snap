// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MainApp dynamically 1`] = `
<div>
  <div>
    <button
      onClick={[Function]}
    >
       Click
    </button>
  </div>
  <div>
    loading...
  </div>
  <button
    onClick={[Function]}
  >
     toggle
  </button>
  <Hello />
  <Memo()
    increment={[Function]}
  />
  <div>
    0
  </div>
  <form
    onSubmit={[Function]}
  >
    <input
      name="text"
      onChange={[Function]}
      value=""
    />
  </form>
  <div
    className="expenses"
  >
    <ListItem
      className="hello"
      description="Hello"
      id={1}
      key="1"
    />
    <ListItem
      className="hello"
      description="Testing"
      id={2}
      key="2"
    />
  </div>
</div>
`;

exports[`test result at class expenses 1`] = `
<div>
  <div>
    <button
      onClick={[Function]}
    >
       Click
    </button>
  </div>
  <div>
    loading...
  </div>
  <button
    onClick={[Function]}
  >
     toggle
  </button>
  <Hello />
  <Memo()
    increment={[Function]}
  />
  <div>
    0
  </div>
  <form
    onSubmit={[Function]}
  >
    <input
      name="text"
      onChange={[Function]}
      value=""
    />
  </form>
  <div
    className="expenses"
  >
    <p>
      Nothing
    </p>
  </div>
</div>
`;
